src/Microsoft.AspNetCore.Dispatcher/DispatcherMiddleware.cs

```diff
@@ -1,60 +1,93 @@
                         {
-                            Endpoint = new DispatcherEndpoint("example"),
-                            RequestDelegate = async (context) =>
+                            var dispatcherFeature = new DispatcherFeature
                             {
-                                await context.Response.WriteAsync("Hello from the example!");
-                            }
+                                Endpoint = endpoint,
+                                RequestDelegate = endpoint.RequestDelegate
+                            };
+
+                            httpContext.Features.Set<IDispatcherFeature>(dispatcherFeature);
                         }
-                },
+                    }
+                }
+            }
+
+            await _next(httpContext);
+        }
+
+        private RouteValueDictionary GetDefaults(RouteTemplate parsedTemplate)
+        {
+            var result = new RouteValueDictionary();
+
+            foreach (var parameter in parsedTemplate.Parameters)
+            {
+                if (parameter.DefaultValue != null)
                 {
-                    "/example2",
-                    new DispatcherFeature
-                        {
-                            Endpoint = new DispatcherEndpoint("example2"),
-                            RequestDelegate = async (context) =>
-                            {
-                                await context.Response.WriteAsync("Hello from the second example!");
-                            }
-                        }
-                },
-            };
+                    result.Add(parameter.Name, parameter.DefaultValue);
+                }
+            }
+
+            return result;
+        }
+
+        private bool MatchURLToEndpoint(RouteValueDictionary currentURLRouteValueDictionary, RouteValueDictionary endpointRouteValueDictionary)
+        {
+            var endpointMatch = new RouteValueDictionary();
 
-            if (dictionary.TryGetValue(httpContext.Request.Path, out var value))
+            foreach (var key in currentURLRouteValueDictionary.Keys)
             {
-                var dispatcherFeature = new DispatcherFeature
+                if (!endpointRouteValueDictionary.Keys.Contains(key))
                 {
-                    Endpoint = value.Endpoint,
-                    RequestDelegate = value.RequestDelegate
-                };
+                    break;
+                }
 
-                httpContext.Features.Set<IDispatcherFeature>(dispatcherFeature);
-                await _next(httpContext);
+                if (endpointRouteValueDictionary[key].Equals(currentURLRouteValueDictionary[key]))
+                {
+                    endpointMatch[key] = endpointRouteValueDictionary[key];
+                }
             }
+
+            if (endpointMatch.Count == currentURLRouteValueDictionary.Count)
+            {
+                return true;
+            }
+
+            return false;
```

Comment:
Suggestion on the algorithm.

Psuedocode
```
for each kvp in requiredValues
{
   if (string.IsNullOrEmpty(kvp.Value))
   {
      if (values.TryGetValue(kvp.Key, out var routeValue) && !string.IsNullOrEmpty(routeValue))
      {
          return false;
      }
   }
   else
   {
        if (!values.TryGetValue(kvp.Key, out var routeValue)  || string.Equals(kvp.Value, routeValue, OrdinalIgnoreCase))
        {
            return false;
         }
    }
}

return true;
```

Fun Facts:
1. All route values need to be converted to strings so they can be compared (usually they already are strings)
2. All route value strings are compared ignoring case
3. Null and empty string are considered equivalent
5. All of the values in the endpoint's route value collection that have a null/empty value require that the corresponding value from the route is null/empty/not-present
6. All of the values in the endpoint's route value collection that have a non-null value require that the corresponding value from the route is equal
7. The route values provided by the route can and will have more values (`.Count`) than the route values required by the endpoint